modules/imgproc/src/resize.cpp

```diff
@@ -3493,6 +3493,43 @@ static bool ocl_resize( InputArray _src, OutputArray _dst, Size dsize,
                    (float)inv_fx, (float)inv_fy);
         }
     }
+    else if (interpolation == INTER_LINEAR_EXACT) {
+        int wdepth = std::max(depth, CV_32S), wtype = CV_MAKETYPE(wdepth, cn);
+
+        char buf[2][50];
+        // Precompute the coefficients and store them in a texture
+        UMat coeffsTex(dsize.height + dsize.width, 2, CV_16SC1);
+        ocl::KernelArg coeffsTexArg = ocl::KernelArg::WriteOnlyNoSize(coeffsTex);
+        k.create("precomputeCoeffs", ocl::imgproc::resize_oclsrc,
+                 format("-D PRECOMPUTE_COEFFS -D depth=%d -D T=%s -D T1=%s "
+                        "-D WT=%s -D convertToWT=%s -D convertToDT=%s -D cn=%d "
+                        "-D INTER_RESIZE_COEF_BITS=%d",
+                        depth, ocl::typeToStr(type), ocl::typeToStr(depth), ocl::typeToStr(wtype),
+                        ocl::convertTypeStr(depth, wdepth, cn, buf[0], sizeof(buf[0])),
+                        ocl::convertTypeStr(wdepth, depth, cn, buf[1], sizeof(buf[1])),
+                        cn, INTER_RESIZE_COEF_BITS));
+        if (k.empty())
+            return false;
```

Comment:
I've added some debug printouts and tried to run the `ResizeLinearExactFixture` perf test, but it falls back to CPU here for some reason (`k` is empty). I tried NVIDIA and Intel OCL runtimes from the Ubuntu 22.04 packages (`intel-opencl-icd`). It doesn't happen with regular `ResizeFixture` test. Do you know what might be the reason?

```
Current OpenCL device: 
  17   │     Type = dGPU
  18   │     Name = NVIDIA GeForce RTX 3060
  19   │     Version = OpenCL 3.0 CUDA
  20   │     Driver version = 545.29.02
```
```
Current OpenCL device: 
  17   │     Type = iGPU
  18   │     Name = Intel(R) UHD Graphics 750 [0x4c8a]
  19   │     Version = OpenCL 3.0 NEO 
  20   │     Driver version = 1.0.0
```

Reply:
Apparently it returns `CL_INVALID_KERNEL_NAME (-46)` result:
```
OpenCV(OpenCL:-46): clCreateKernel('precomputeCoeffs')
```

Reply:
OK, I see, you used `-D PRECOMPUTE_COEFFS` but the function is guarded by `-D INTER_LINEAR_EXACT` macro.

Reply:
With this macro enabled we get some OCL compile errors (line numbers are broken):
```
OpenCV(OpenCL:0): clBuildProgram(source: -D INTER_LINEAR_EXACT -D depth=0 -D T=uchar -D T1=uchar -D WT=int -D convertToWT=convert_int -D convertToDT=convert_uchar_sat -D cn=1 -D INTER_RESIZE_COEF_BITS=11)
OpenCL program build log: imgproc/resize
Status -11: CL_BUILD_PROGRAM_FAILURE
-D INTER_LINEAR_EXACT -D depth=0 -D T=uchar -D T1=uchar -D WT=int -D convertToWT=convert_int -D convertToDT=convert_uchar_sat -D cn=1 -D INTER_RESIZE_COEF_BITS=11
<kernel>:165:46: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
ialpha = convert_short_sat_rte((1.f - fxx) * INTER_RESIZE_COEF_SCALE);
                                             ^
<kernel>:166:45: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
ibeta = convert_short_sat_rte((1.f - fyy) * INTER_RESIZE_COEF_SCALE);
                                            ^
<kernel>:179:11: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
ialpha1 = INTER_RESIZE_COEF_SCALE - ialpha0;
          ^
<kernel>:182:10: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
ibeta1 = INTER_RESIZE_COEF_SCALE - ibeta0;
         ^
<kernel>:185:12: error: use of undeclared identifier 'src_stride'
sy0 = sy * src_stride;
           ^
<kernel>:186:13: error: use of undeclared identifier 'src_stride'
sy1 = sy0 + src_stride;
            ^
<kernel>:197:24: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
fres = fres * (1.0f / (INTER_RESIZE_COEF_SCALE * INTER_RESIZE_COEF_SCALE));
                       ^
<kernel>:197:50: error: use of undeclared identifier 'INTER_RESIZE_COEF_SCALE'
fres = fres * (1.0f / (INTER_RESIZE_COEF_SCALE * INTER_RESIZE_COEF_SCALE));
                                                 ^
<kernel>:199:31: error: use of undeclared identifier 'dst_offset'
write_imagei(dst, (int2)(dx + dst_offset, dy + dst_offset), res);
                              ^

```

Reply:
I have a suspicion... could you please provide cl_info for both devices?

Reply:
I just had a look and found a mess of a last commit. i wondered why that happened and it turn out it was one of my experiments and that it landed in my commit because a project build with this cmake configuration...
```bash
cmake -DINSTALL_BIN_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_OPENCL=ON -DBUILD_opencv_java=OFF -DBUILD_opencv_js=OFF -DBUILD_opencv_python2=OFF -DBUILD_opencv_python3=OFF -DBUILD_EXAMPLES=OFF -DBUILD_PACKAGE=OFF -DBUILD_TESTS=ON -DBUILD_PERF_TESTS=ON -DBUILD_DOCS=OFF -DBUILD_opencv_plot=OFF -DBUILD_opencv_tracking=OFF -DBUILD_opencv_video=OFF -DBUILD_opencv_world=OFF ../opencv
```
acts like it rebuilds required files after updating resize.cl like this...
```bash
[ 56%] Processing OpenCL kernels (imgproc)
-- /home/elchaschab/devel/opencv_build/modules/imgproc/opencl_kernels_imgproc.hpp contains the same content
[ 56%] Building CXX object modules/imgproc/CMakeFiles/opencv_imgproc.dir/opencl_kernels_imgproc.cpp.o
[ 57%] Linking CXX shared library ../../lib/libopencv_imgproc.so
[100%] Built target opencv_imgproc
```

but when running opencv_perf_imgproc it still uses the old resize.cl. I still need to find out why, because simply deleting resize.cpp.o and opencv_perf_imgproc doesn't fix it. the get rebuild but still the old cl code. only thing that helps is ```make clean```

Reply:
had to update the cmake line, but still the same problem:
```bash
cmake -DINSTALL_BIN_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_OPENCL=ON -DBUILD_opencv_java=OFF -DBUILD_opencv_js=OFF -DBUILD_opencv_python2=OFF -DBUILD_opencv_python3=OFF -DBUILD_EXAMPLES=OFF -DBUILD_PACKAGE=OFF -DBUILD_TESTS=ON -DBUILD_PERF_TESTS=ON -DBUILD_DOCS=OFF -DBUILD_opencv_videoio=ON -DBUILD_opencv_highgui=ON -DBUILD_opencv_ts=ON -DBUILD_opencv_imgcodecs=ON -DBUILD_opencv_plot=OFF -DBUILD_opencv_tracking=OFF -DBUILD_opencv_video=OFF -DBUILD_opencv_world=OFF ../opencv
```

Reply:
OpenCV caches compiled OpenCL kernels in `~/.cache/opencv/` folder, try to clean it. Perhaps Intel OCL runtime might cache kernels too, but I don't remember details.

Reply:
Well. I threw the build directory away and cleared caches. now it's back to working. for now i will roll back one commit. and retest my attempts.